src/Component/RuleTable/RuleTable.tsx

```diff
@@ -0,0 +1,297 @@
+}
+
+export interface RuleRecord extends GsRule {
+  key: number;
+}
+
+export class RuleTable extends React.Component<RuleTableProps, RuleTableState> {
+
+  static componentName: string = 'RuleTable';
+
+  constructor(props: RuleTableProps) {
+    super(props);
+    this.state = {
+      ruleEditIndex: null,
+      symbolizerEditorVisible: false,
+      filterEditorVisible: false
+    };
+  }
+
+  public static defaultProps: RuleTableDefaultProps = {
+    locale: en_US.GsRuleTable
+  };
+
+  public shouldComponentUpdate(nextProps: RuleTableProps, nextState: RuleTableState): boolean {
+    const diffProps = !_isEqual(this.props, nextProps);
+    const diffState = !_isEqual(this.state, nextState);
+    return diffProps || diffState;
+  }
+
+  getRuleRecords = (): RuleRecord[] => {
+    const {
+      rules
+    } = this.props;
+    return rules.map((rule: GsRule, index: number): RuleRecord => {
+      return {
+        key: index,
+        ...rule
+      };
+    });
+  }
+
+  onSymbolizerClick = (record: RuleRecord) => {
+    this.setState({
+      ruleEditIndex: record.key,
+      symbolizerEditorVisible: true,
+      filterEditorVisible: false
+    });
+  }
+
+  symbolizerRenderer = (text: string, record: RuleRecord, index: number) => {
+    return (
+      <Renderer
+        symbolizers={record.symbolizers}
+        onClick={() => {
+          this.onSymbolizerClick(record);
+        }}
+      />
+    );
+  }
+
+  nameRenderer = (text: string, record: RuleRecord, index: number) => {
+    return (
+      <Input
+        value={record.name}
+        onChange={(event) => {
+          const newName = event.target.value;
+          this.setValueForRule(record.key, 'name', newName);
+        }}
+      />
+    );
+  }
+
+  filterRenderer = (text: string, record: RuleRecord, index: number) => {
+    return (
+      <Input.Search
+        value={JSON.stringify(record.filter)}
+        onChange={(event) => {
+          try {
+            const newFilter = JSON.parse(event.target.value);
+            this.setValueForRule(record.key, 'filter', newFilter);
+          } catch (error) {
+            // TODO Feedback
+          }
+        }}
+        enterButton={(
+          <Icon
+            type="edit"
+          />
+        )}
+        onSearch={() => {
+          this.onFilterEditClick(record.key);
+        }}
+      />
+    );
+  }
+
+  onFilterEditClick = (ruleEditIndex: number) => {
+    this.setState({
+      ruleEditIndex,
+      symbolizerEditorVisible: false,
+      filterEditorVisible: true
+    });
+  }
+
+  minScaleRenderer = (text: string, record: RuleRecord, index: number) => {
+    const minScaleDenominator = _get(record, 'scaleDenominator.min');
+    const value = minScaleDenominator ? parseFloat(minScaleDenominator) : undefined;
+    return (
+      <InputNumber
+        value={value}
+        min={0}
+        formatter={val => `1:${val}`}
+        parser={(val: string) => parseFloat(val.replace('1:', ''))}
```

Comment:
You could use the property `addonBefore` instead `formatter` and `parser` here. This adds a label before the input, like this:

![image](https://user-images.githubusercontent.com/12186477/48069737-76f7c380-e1d6-11e8-82dc-234182abdcb9.png)

Then you would not have to format and parse the scales and empty inputs would still look fine.

Reply:
Yeah i saw that. Still not sure what looks better...